codesyncapp/codesync-vs
# S3 uploader

**S3 Uploader**

	We'll have YML files in `~/.codesync/.s3_uploader/`
	Each file will look like
	
		repo_path: /Users/dev/codesync/codesync-vs
		branch: master
		file_path_and_urls:
			file_path_1: url_1
			file_path_2: url_2
			...
		locked_by: uuid
		failed_count: 0
	
**Flow**
- For valid paths and URLs, we create tasks to upload each file to s3.
- We then iterate chunks of those tasks to slow the upload process a bit. 
	e.g. we have 450 files then with default chunk size of 100, we will have 5 chunks.
	
**Error Handling**
- If some chunk is failed to upload successfully somehow, we save that chunk in a separate file to retry later.
- To retry a failed chunk, we reduce the chunk size to 50 to try with less concurrent uploads.
- We repeat this process 3 times, finally with the chunk size of 25.
- If failed count becomes 3 for a file, we ignore that file for now.
	
**Internet Outage**
- If internet is offline, it retires after 5 minutes.

**Testing Instructions**
1. Try to create new branch (e.g. in webapp) and make sure all chunks are uploaded correctly on s3. 
2. If any chunk is failed, it should be retried after iterating remaining chunks. Retrying chunk should have smaller size than DEFAULT CHUNK SIZE which is 100
3. If internet is down, s3 uploader should not work. Files in .originals must not be deleted. s3Uploader should retry after 5 minutes
